### ISSUE TITLE
Incorrectly generated roof (maybe due to floating point errors)

### ISSUE THREAD

**Issue by costika1234** [prochitecture/bpypolyskel#13]
I've been able to find an example where `bpypolyskel` fails to generate the correct hip roof for a simple square base with a square hole right in the middle. Given the following code

```
import math
import mathutils
import matplotlib.pyplot as plt

from bpypolyskel import bpypolyskel
from matplotlib.widgets import Slider


_COORDS = [
    # Outer loop (CCW).
    [78.6591, 430.332, 0],
    [78.6591, 76.0013, 0],
    [432.99, 76.0013, 0],
    [432.99, 430.332, 0],
    # Inner loop (CW).
    [196.769, 194.112, 0],
    [196.769, 312.222, 0],
    [314.88, 312.222, 0],
    [314.88, 194.112, 0],
]

VERTS = [mathutils.Vector(coords) for coords in _COORDS]

FIRST_VERTEX_INDEX = 0
NUM_VERTS = 4
HOLES_INFO = [(4, 4)]

FACES, UNIT_VECTORS = None, None


def create_roof(pitch, ax):
    ax.clear()

    faces = bpypolyskel.polygonize(
        VERTS,
        FIRST_VERTEX_INDEX,
        NUM_VERTS,
        HOLES_INFO,
        0.0,
        math.tan(pitch * math.pi / 180),
        FACES,
        UNIT_VECTORS,
    )

    # Plot the hipped roof in 3D.
    for face in faces:
        for edge in zip(face, face[1:] + face[:1]):
            p1, p2 = VERTS[edge[0]], VERTS[edge[1]]
            ax.plot([p1.x, p2.x], [p1.y, p2.y], [p1.z, p2.z], 'k')

    ax.axis('equal')
    ax.set_zlim(0, 450)


def run_demo():
    fig = plt.figure('Roof Demo')
    ax = fig.add_subplot(projection='3d')

    # Display the roof using the default pitch.
    create_roof(30.0, ax)

    plt.show()

if __name__ == "__main__":
    # Execute only if run as a script.
    run_demo()
```

the output looks something like this: 
<img width="752" alt="Screen Shot 2023-03-14 at 11 05 16" src="https://user-images.githubusercontent.com/7645454/224950942-e3263a4d-66e9-464c-ac54-839c5178934b.png">

If I change the two `76.0013` values from the `_COORDS` list to `76.00`, i.e. 
```
_COORDS = [
    # Outer loop (CCW).
    [78.6591, 430.332, 0],
    [78.6591, 76.00, 0],
    [432.99, 76.00, 0],
    [432.99, 430.332, 0],
    # Inner loop (CW).
    [196.769, 194.112, 0],
    [196.769, 312.222, 0],
    [314.88, 312.222, 0],
    [314.88, 194.112, 0],
]
```
then the correct roof is generated:
<img width="752" alt="Screen Shot 2023-03-14 at 11 05 29" src="https://user-images.githubusercontent.com/7645454/224951353-9d3cffd6-ef51-4fb1-bb35-d72a1ac4db4c.png">

It definitely looks like floating point errors crept in and they prevent the program from generating the desired hip roof. I was thinking to mitigate this issue by supplying coordinates with at most 2 decimal places, even though this problem might still manifest occasionally. Any other ideas?

@vvoovv @polarkernel

**Comment by polarkernel** [comment]
@costika1234 Thanks for this example. I am almost certain that floating point issues are responsible in this case. I have occasionally observed similar problems with very symmetric footprints. In our application we used footprints from OpenStreetMap building data, where the operator often introduces small deviations from symmetry by small  errors. Perhaps adding small random numbers to your coordinates might help.

**Comment by polarkernel** [comment]
Sorry for closing, was the wrong button.

**Comment by costika1234** [comment]
That's okay, no worries.

Indeed, symmetric footprints can cause the program to generate incorrect roofs, but I've noticed that this only applies to those footprints that contain holes right in the middle. Otherwise, the library performs incredibly well.

Given that I'm processing the coordinates inside my SketchUp plugin, I've resorted to rounding the x and y coordinates of all inner loops to the first decimal place. I was unable to replicate this bug after doing that, so I consider this a good enough solution for the time being.

I guess we can actually close this issue, while admitting that inputs like the above may cause the program to behave incorrectly.
